src/keywrap/pkcs11/mod.rs

```diff
@@ -0,0 +1,333 @@
+use anyhow::{anyhow, Result};
+use std::collections::HashMap;
+use pkcs11_uri::{Pkcs11Uri};
+use crate::keywrap::KeyWrapper;
+use crate::config::{CryptoConfig, EncryptConfig, DecryptConfig, Pkcs11Config,
+                    get_default_module_directories_yaml,
+                    parse_pkcs11_config_file};
+use crate::softhsm::{SoftHSMSetup};
+use crate::utils::{parse_public_key, parse_private_key,
+                   encrypt_multiple, decrypt_pkcs11, KeyType};
+use rsa::{RsaPublicKey};
+use crate::pkcs11_uri_wrapped::Pkcs11UriWrapped;
+
+
+
+#[derive(Debug)]
+pub struct Pkcs11KeyWrapper {}
+
+
+
+// Pkcs11KeyFileObject is a representation of the Pkcs11KeyFile with the pkcs11
+// URI wrapper as an object
+pub struct Pkcs11KeyFileObject {
+    pub uriw: Pkcs11UriWrapped,
+}
+
+
+
+
+impl KeyWrapper for Pkcs11KeyWrapper {
+    
+    // Wrap the session key for recpients and encrypt the opts_data,
+    // which describe the symmetric key used for encrypting the layer
+    fn wrap_keys(&self,
+                 ec: &EncryptConfig,
+                 opts_data: &[u8],
+    ) -> Result<Vec<u8>> {
+        let mut pubkeys: Vec<Vec<u8>> = Vec::new();
+        match ec.param.get("pkcs11-pubkeys") {
+            Some(pks) => {
+                for p in pks {
+                    pubkeys.push(p.to_vec());
+                }
+            },
+            None => (),
+        }
+        match ec.param.get("pkcs11-yamls") {
+            Some(yamls) => {
+                for y in yamls {
+                    pubkeys.push(y.to_vec());
+                }
+            },
+            None => (),
+        };
+        let dc = match ec.decrypt_config.as_ref() {
+            Some(pubkeys) => pubkeys,
+            None => return Err(anyhow!("EncryptConfig is missing
+                                        decrypt_config member")),
+        };
+
+        let pkcs11_recipients: Vec<KeyType>
+            = add_pub_keys(&dc, &pubkeys)?;
+
+        if pkcs11_recipients.is_empty() {
+            return Ok(Vec::new())
+        }
+
+        let json_str = encrypt_multiple(&pkcs11_recipients, opts_data)?;
+
+        Ok(json_str)
+    }
+
+    fn unwrap_keys(&self,
+                   dc: &DecryptConfig,
+                   annotation: &[u8],
+    ) -> Result<Vec<u8>> {
+
+        let mut pkcs11_keys = Vec::new();
+
+        let priv_keys = match self.private_keys(&dc.param) {
+            Some(x) => x,
+            None => return Err(anyhow!("")),
+        };
+
+        let p11conf_opt = p11conf_from_params(&dc.param)?;
+
+        for key in priv_keys {
+
+            let mut k = parse_private_key(&key, &vec![], "PKCS11".to_string())?;
+            match k {
```

Comment:
This match can also be an `if let`, I believe, which would look cleaner.
Also, we only use `k` when `p11conf_opt` has `Some` value. That check could be done outside this entire loop, presumably. But then that assumes `parse_private_key()` has no side effects, so check that. And finally, this begs the question Dov had below: Do we need to check that pkcs11_keys is non-empty?